src/hooks/core/users/useFollowUserV2Hook.tsx

```diff
@@ -0,0 +1,135 @@
+
+    try {
+      const result = await checkFollowStatus(channelId);
+      StorageUtils.otherProfileData.setForKey(channelId, JSON.stringify(result));
+    } catch (e) {
+      console.error('get follow status failed', e);
+    }
+  }
+
+  const getOtherProfileFollowStatus = async (channelId: string | undefined): Promise<any> => {
+    if (!channelId) return;
+
+    let otherProfileDataCache = StorageUtils.otherProfileData.getForKey(channelId);
+    let otherProfileDataJSON: FollowStatusAPI = {
+      isMeFollowingTargetFromAction: undefined,
+      isMeFollowingTarget: undefined,
+      isTargetFollowingMe: undefined
+    };
+
+    try {
+      otherProfileDataJSON = JSON.parse(otherProfileDataCache || '{}');
+    } catch (e) {
+      console.error(e);
+    }
+
+    if (otherProfileDataCache) {
+      setFollowStatus({
+        isFollowed: otherProfileDataJSON?.isTargetFollowingMe,
+        isFollowing: otherProfileDataJSON?.isMeFollowingTarget,
+        isFollowingFromAction: otherProfileDataJSON?.isMeFollowingTargetFromAction
+      });
+    }
+
+    await __helperCallOtherProfileData(channelId);
+    otherProfileDataCache = StorageUtils.otherProfileData.getForKey(channelId);
+    try {
+      otherProfileDataJSON = JSON.parse(otherProfileDataCache || '{}');
+    } catch (e) {
+      console.error(e);
+    }
+
+    setFollowStatus({
+      isFollowed: otherProfileDataJSON?.isTargetFollowingMe,
+      isFollowing: otherProfileDataJSON?.isMeFollowingTarget,
+      isFollowingFromAction: otherProfileDataJSON?.isMeFollowingTargetFromAction
+    });
+  };
+
+  const followUserAction = async (item: ChannelList) => {
+    /**
+     * Optimistic state update (START)
+     */
+    const initialFollowStatus = {...followStatus};
+    setFollowStatus((prevFollowStatus) => ({
+      ...prevFollowStatus,
+      isFollowingFromAction: !initialFollowStatus?.isFollowingFromAction,
+      isFollowing: !initialFollowStatus?.isFollowing
+    }));
+
+    /**
+     * Optimistic state update (END)
+     */
+
+    try {
+      const myId = tab === 'signed' ? signedProfileId : anonProfileId;
+      const targetUser = item?.rawJson?.channel?.members?.find(
+        (member) => member?.user_id !== myId
+      )?.user;
+
+      const data = {
+        user_id_followed: targetUser?.id,
+        follow_source: 'chat'
+      };
+
+      if (initialFollowStatus?.isFollowing) {
+        await setUnFollow(data);
+      } else {
+        await setFollow(data);
+      }
+      __helperCallOtherProfileData(item?.id);
+    } catch (e) {
+      const action = initialFollowStatus?.isFollowing ? 'unfollow' : 'follow';
+      SimpleToast.show(`Failed to ${action} user. Please try again later`);
+      /**
+       * Revert optimistic update
+       */
+      setFollowStatus(() => ({
+        ...initialFollowStatus
+      }));
+    }
+  };
+
+  return {
+    followStatus,
+
+    followUserAction,
+    getOtherProfileFollowStatus
+  };
+};
+
+export default useFollowUserV2Hook;
```

Comment:
<img src="https://avatars.githubusercontent.com/in/347564?s=41" alt="Image description" width="20" height="20">   CodeRabbit

The `useFollowUserV2Hook` function seems to be well implemented. It uses the React Hook `useState` to manage the follow status and provides functions to get the follow status and perform follow/unfollow actions. The hook also handles optimistic UI updates by immediately updating the state before making API calls, and then reverting the update if the API call fails. This can provide a smoother user experience as it reduces perceived latency.

However, there are some areas that could be improved:

1. Error handling: Currently, when an error occurs in the `__helperCallOtherProfileData` or `getOtherProfileFollowStatus` functions, it is logged to the console but not propagated further. This might make it difficult to handle these errors at a higher level. Consider throwing the error after logging it, or returning a specific error value that can be checked by the caller.

2. AsyncStorage usage: The `StorageUtils.otherProfileData.getForKey` and `setForKey` methods are used synchronously, which might block the JavaScript thread if the stored data is large or the device is slow. Consider using the asynchronous versions of these methods instead.

3. Redundant code: The code for parsing the cached profile data and setting the follow status is duplicated in the `getOtherProfileFollowStatus` function. Consider extracting this into a separate function to adhere to the DRY (Don't Repeat Yourself) principle.

4. Type safety: The `getOtherProfileFollowStatus` function is declared to return a Promise, but it doesn't always return a value. This could lead to unexpected behavior. Consider always returning a value, even in error cases.

Here are the suggested changes:

```diff
  async function __helperCallOtherProfileData(channelId: string) {
    if (!channelId) throw new Error('Channel ID is required');

    try {
      const result = await checkFollowStatus(channelId);
-     StorageUtils.otherProfileData.setForKey(channelId, JSON.stringify(result));
+     await StorageUtils.otherProfileData.setForKeyAsync(channelId, JSON.stringify(result));
    } catch (e) {
      console.error('get follow status failed', e);
+     throw e;
    }
  }

  const getOtherProfileFollowStatus = async (channelId: string | undefined): Promise<any> => {
    if (!channelId) throw new Error('Channel ID is required');

-   let otherProfileDataCache = StorageUtils.otherProfileData.getForKey(channelId);
+   let otherProfileDataCache = await StorageUtils.otherProfileData.getForKeyAsync(channelId);
    let otherProfileDataJSON: FollowStatusAPI = {
      isMeFollowingTargetFromAction: undefined,
      isMeFollowingTarget: undefined,
      isTargetFollowingMe: undefined
    };

    try {
      otherProfileDataJSON = JSON.parse(otherProfileDataCache || '{}');
    } catch (e) {
      console.error(e);
+     throw e;
    }

    setFollowStatusFromCache(otherProfileDataJSON);

    await __helperCallOtherProfileData(channelId);
-   otherProfileDataCache = StorageUtils.otherProfileData.getForKey(channelId);
+   otherProfileDataCache = await StorageUtils.otherProfileData.getForKeyAsync(channelId);
    try {
      otherProfileDataJSON = JSON.parse(otherProfileDataCache || '{}');
    } catch (e) {
      console.error(e);
+     throw e;
    }

-   setFollowStatus({
-     isFollowed: otherProfileDataJSON?.isTargetFollowingMe,
-     isFollowing: otherProfileDataJSON?.isMeFollowingTarget,
-     isFollowingFromAction: otherProfileDataJSON?.isMeFollowingTargetFromAction
-   });
+   setFollowStatusFromCache(otherProfileDataJSON);
+   return otherProfileDataJSON;
  };

+ const setFollowStatusFromCache = (otherProfileDataJSON: FollowStatusAPI) => {
+   setFollowStatus({
+     isFollowed: otherProfileDataJSON?.isTargetFollowingMe,
+     isFollowing: otherProfileDataJSON?.isMeFollowingTarget,
+     isFollowingFromAction: otherProfileDataJSON?.isMeFollowingTargetFromAction
+   });
+ };
```


<!-- This is an auto-generated comment by OSS CodeRabbit -->